<a href="https://colab.research.google.com/github/msheikh1/Week-5-team-project/blob/main/Week_5_San_Francisco_Salary_Team_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Data Science Prep Final Week Project**
Team Members:
* Masud sheikh
* Halima I Hussein
* Joan Mwangi
* Nathan Njonge
* Sammy Mwangi

## **San Francisco Salary Analysis for Years 2011 to 2014**

###**Problem Statement**

Irrespective of professional experience and educational background, gender pay disparity is a problem in the public sector. Women have to overcome salary barriers, such as agency segregation, position segregation. Recent studies have indicated that human capital variables, people skills, discrimination, and policies all contribute to gender pay disparity in America’s workforce. However, there are limited studies that focus on the indirect factors that also contribute to gender pay inequality.
The purpose of this study was to investigate the relationship between wages and job responsibility as well as gender (as defined by an employee’s job title) for all employees working in the state of San Francisco on an annual basis from 2011 to 2014. The data source for this study came from Kaggle website. [Link](https://www.kaggle.com/ronaldtroncoso20/sf-salaries-extended/download) 
 


###**Objectives**

Our objectives were:
* To identify how salaries changed over time between different groups of people
* To identify how far the male salary and female salary deviate from mean salary
* Identify how base pay, overtime pay, and benefits are allocated between different groups
* Identify any evidence of pay discrimination based on gender in this dataset
* Identify the difference in  wages and gender and an employee’s type of job

#Data Preparation

##Importing Libraries

In [ ]:
import pandas as pd
import numpy as np

##Importing the dataset

In [ ]:
# We have downloaded the dataset from the below url
#url = 'https://www.kaggle.com/ronaldtroncoso20/sf-salaries-extended/download'
#
df = pd.read_csv('SF_salary_data_gender.csv')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (4,5,6,7,13) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


##Previewing the dataset

In [ ]:
# dataframe info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 15 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Unnamed: 0        148654 non-null  int64  
 1   Id                148654 non-null  int64  
 2   Employee Name     148653 non-null  object 
 3   JobTitle          148654 non-null  object 
 4   BasePay           148049 non-null  object 
 5   OvertimePay       148654 non-null  object 
 6   OtherPay          148654 non-null  object 
 7   Benefits          112495 non-null  object 
 8   TotalPay          148654 non-null  float64
 9   TotalPayBenefits  148654 non-null  float64
 10  Year              148654 non-null  int64  
 11  Notes             0 non-null       float64
 12  Agency            148654 non-null  object 
 13  Status            38119 non-null   object 
 14  gender            119169 non-null  object 
dtypes: float64(3), int64(3), object(9)
memory usage: 17.0+ MB


In [ ]:
#Previewing the first 10 rows of the dataset
df.head(10)

Unnamed: 0  Id       Employee Name  ...         Agency Status gender
0           0   1      nathaniel ford  ...  San Francisco    NaN      m
1           1   2        gary jimenez  ...  San Francisco    NaN      m
2           2   3      albert pardini  ...  San Francisco    NaN      m
3           3   4   christopher chong  ...  San Francisco    NaN      m
4           4   5     patrick gardner  ...  San Francisco    NaN      m
5           5   6      david sullivan  ...  San Francisco    NaN      m
6           6   7           alson lee  ...  San Francisco    NaN    NaN
7           7   8       david kushner  ...  San Francisco    NaN      m
8           8   9      michael morris  ...  San Francisco    NaN      m
9           9  10  joanne hayes-white  ...  San Francisco    NaN      f

[10 rows x 15 columns]

In [ ]:
# Viewing the data description
#we notice that there are negative values
df.describe()

Unnamed: 0             Id  ...           Year  Notes
count  148654.000000  148654.000000  ...  148654.000000    0.0
mean    74326.500000   74327.500000  ...    2012.522643    NaN
std     42912.857795   42912.857795  ...       1.117538    NaN
min         0.000000       1.000000  ...    2011.000000    NaN
25%     37163.250000   37164.250000  ...    2012.000000    NaN
50%     74326.500000   74327.500000  ...    2013.000000    NaN
75%    111489.750000  111490.750000  ...    2014.000000    NaN
max    148653.000000  148654.000000  ...    2014.000000    NaN

[8 rows x 6 columns]

##Cleaning the data

###Validity

In [ ]:
# drop columns
df.drop(['Unnamed: 0', 'Notes', 'Status', 'Id'], axis = 1, inplace = True)

In [ ]:
#confirming the remaining columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Employee Name     148653 non-null  object 
 1   JobTitle          148654 non-null  object 
 2   BasePay           148049 non-null  object 
 3   OvertimePay       148654 non-null  object 
 4   OtherPay          148654 non-null  object 
 5   Benefits          112495 non-null  object 
 6   TotalPay          148654 non-null  float64
 7   TotalPayBenefits  148654 non-null  float64
 8   Year              148654 non-null  int64  
 9   Agency            148654 non-null  object 
 10  gender            119169 non-null  object 
dtypes: float64(2), int64(1), object(8)
memory usage: 12.5+ MB


In [ ]:
#Drop rows where Base Pay is indicated not provided
df.drop(df.index[df['BasePay'] == 'Not Provided'], inplace = True)

###Uniformity

In [ ]:
#convert datatypes
# strings to float
df["BasePay"] = df["BasePay"].astype(float)
df["OvertimePay"] = df["OvertimePay"].astype(float)
df["OtherPay"] = df["OtherPay"].astype(float)
df["Benefits"] = df["Benefits"].astype(float)
df.dtypes

Employee Name        object
JobTitle             object
BasePay             float64
OvertimePay         float64
OtherPay            float64
Benefits            float64
TotalPay            float64
TotalPayBenefits    float64
Year                  int64
Agency               object
gender               object
dtype: object

In [ ]:
#Changing column names to be uniform
df.columns =['EmployeeName',	'JobTitle',	'BasePay',	'OvertimePay',	'OtherPay',	'Benefits',	'TotalPay',	'TotalPayBenefits',	'Year',	'Agency',	'Gender']


In [ ]:
#Dropping rows with negative values
df.drop(df.index[df['TotalPay'] < 0], inplace = True)
df.drop(df.index[df['BasePay'] < 0], inplace = True)
df.drop(df.index[df['OvertimePay'] < 0], inplace = True)
df.drop(df.index[df['OtherPay'] < 0], inplace = True)
df.drop(df.index[df['Benefits']< 0], inplace = True)
df.drop(df.index[df['BasePay'] < 0], inplace = True)
df.describe()

BasePay    OvertimePay  ...  TotalPayBenefits           Year
count  148027.000000  148629.000000  ...     148629.000000  148629.000000
mean    66331.428168    5066.750620  ...      93705.402123    2012.522597
std     42762.323488   11455.040983  ...      62789.023704       1.117578
min         0.000000       0.000000  ...          0.000000    2011.000000
25%     33599.910000       0.000000  ...      44106.370000    2012.000000
50%     65010.820000       0.000000  ...      92412.080000    2013.000000
75%     94691.060000    4658.830000  ...     132882.060000    2014.000000
max    319275.010000  245131.880000  ...     567595.430000    2014.000000

[8 rows x 7 columns]

###Completeness

In [ ]:
#Checking for missing values
df.isnull().any()

EmployeeName         True
JobTitle            False
BasePay              True
OvertimePay         False
OtherPay            False
Benefits             True
TotalPay            False
TotalPayBenefits    False
Year                False
Agency              False
Gender               True
dtype: bool

In [ ]:
# Change null values to 0 in BasePay, OvertimePay, OtherPay and Benefits columns
df.BasePay=df.BasePay.fillna(0)
df["OvertimePay"] = df["OvertimePay"].fillna(0)
df["OtherPay"] = df["OtherPay"].fillna(0)
df["Benefits"] = df["Benefits"].fillna(0)

###Consistency

In [ ]:
# Checking for duplicates
df.duplicated().any()

True

In [ ]:
#Dropping the duplicates
df = df.drop_duplicates()

###Accuracy

In [ ]:
#Checking whether BasePay, OvertimePay, OtherPay and Benefits totals to TotalPayBenefits  
# Displaying those that do not tally
df[df.BasePay + df.OvertimePay + df.OtherPay + df.Benefits != df.TotalPayBenefits]


EmployeeName  ... Gender
0          nathaniel ford  ...      m
7           david kushner  ...      m
10          arthur kenney  ...      m
13            john martin  ...      m
15        richard corriea  ...      m
...                   ...  ...    ...
148582              eva m  ...      f
148596           rowena t  ...    NaN
148600              kim l  ...      f
148601  brieanna mcfadden  ...    NaN
148604            sugyn m  ...    NaN

[36094 rows x 11 columns]

In [ ]:
# Since we have rows where BasePay, OvertimePay, OtherPay and Benefits does not totals to TotalPayBenefits,
# We create a new colum which is a sum of BasePay,	OvertimePay,	OtherPay and	Benefits and call it 'Pay'
# 

df['Pay'] = df.BasePay + df.OvertimePay + df.OtherPay +df.Benefits

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 148626 entries, 0 to 148649
Data columns (total 12 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   EmployeeName      148625 non-null  object 
 1   JobTitle          148626 non-null  object 
 2   BasePay           148626 non-null  float64
 3   OvertimePay       148626 non-null  float64
 4   OtherPay          148626 non-null  float64
 5   Benefits          148626 non-null  float64
 6   TotalPay          148626 non-null  float64
 7   TotalPayBenefits  148626 non-null  float64
 8   Year              148626 non-null  int64  
 9   Agency            148626 non-null  object 
 10  Gender            119145 non-null  object 
 11  Pay               148626 non-null  float64
dtypes: float64(7), int64(1), object(4)
memory usage: 14.7+ MB


In [ ]:
#We now check whether the columns tally by displaying any row that does not tally
df[df.BasePay + df.OvertimePay + df.OtherPay + df.Benefits != df.Pay]

Empty DataFrame
Columns: [EmployeeName, JobTitle, BasePay, OvertimePay, OtherPay, Benefits, TotalPay, TotalPayBenefits, Year, Agency, Gender, Pay]
Index: []

###Outliers

In [ ]:
#We use 5th and 95 percentiles to see whether there are outliers
# 
Q1 = df.quantile(0.05)
Q3 = df.quantile(0.95)

IQR = Q3 - Q1

outliers=(df < (Q1 - 1.5 * IQR)) |(df > (Q3 + 1.5 * IQR))
outliers.head()

Agency  BasePay  Benefits  ...  TotalPay  TotalPayBenefits   Year
0   False    False     False  ...      True              True  False
1   False    False     False  ...      True              True  False
2   False    False     False  ...     False             False  False
3   False    False     False  ...     False             False  False
4   False    False     False  ...     False             False  False

[5 rows x 12 columns]

In [ ]:
#We also plot box plots to see the extent of the outliers
import matplotlib.pyplot as plt 
% matplotlib inline 

outliers=[]

def detect_outlier(df):
    
    threshold=3
    mean_1 = np.mean(df)
    std_1 =np.std(df)
    
    for y in df:
        z_score= (y - mean_1)/std_1 
        if np.abs(z_score) > threshold:
            outliers.append(y)
    return outliers

detect_outlier(df.TotalPay)

[567595.43,
 538909.28,
 335279.91,
 332343.61,
 326373.19,
 316285.74,
 315981.05,
 307899.46,
 303427.55,
 302377.73,
 299494.17,
 297608.92,
 294580.02,
 292671.62,
 286347.05,
 286213.86,
 284720.43,
 278569.21,
 276434.22,
 274550.25,
 274190.27,
 273771.21,
 273702.71,
 270672.63,
 270324.91,
 267992.59,
 265784.56,
 265463.46,
 264074.6,
 261366.14,
 260297.66,
 259927.11,
 259684.88,
 258588.39,
 258364.22,
 257540.09,
 257124.44,
 253269.44,
 251935.01,
 251685.16,
 250595.39,
 250408.49,
 250339.35,
 250317.93,
 250122.04,
 249375.02,
 249293.3,
 248895.77,
 248718.64,
 248434.39,
 247254.61,
 246305.62,
 244687.81,
 242742.85,
 241286.42,
 241211.17,
 240846.3,
 240019.1,
 239426.36,
 238879.2,
 238803.56,
 238551.88,
 238501.05,
 238479.28,
 237837.73,
 237708.83,
 237344.41,
 236833.82,
 236200.92,
 234865.2,
 234848.11,
 233944.48,
 233464.63,
 233357.28,
 232702.58,
 232486.72,
 232007.77,
 231712.8,
 231652.13,
 231592.25,
 231306.04,
 231104.97,
 230906.51,
 230827.12,

##Exporting our data

In [ ]:
#we export the data to be used by every team member for analysis
df.to_csv('SF_Salary.csv')

#Data analysis

###Importing libraries and dataset

In [ ]:
#Importing our libraries pandas and numpy
import pandas as pd

import numpy as np


In [ ]:
#Importing our csv file
df = pd.read_csv("SF_Salary.csv")
#Getting a preview with 3 rows
df.head(3)



Unnamed: 0    EmployeeName  ... Gender        Pay
0           0  nathaniel ford  ...      m  567595.43
1           1    gary jimenez  ...      m  538909.28
2           2  albert pardini  ...      m  335279.91

[3 rows x 13 columns]

In [ ]:
#Deleting the unnamed column
del df["Unnamed: 0"]


##Highest paid employees

In [ ]:
#Highest paid employees with their salaries
#In this case we grouped by employee name
#We then we got the max pay of the employees to get the name and salary
df.groupby("EmployeeName").Pay.max().nlargest(5)

EmployeeName
nathaniel ford    567595.43
gary jimenez      538909.28
david shinn       510732.68
amy p             479652.21
william j         436224.36
Name: Pay, dtype: float64

###Highest paid females

In [ ]:
#Highest paid females
df[(df.Gender =="f")].groupby("EmployeeName").Pay.max().nlargest(5)

EmployeeName
amy p                 479652.21
joanne m              422353.40
ellen g               415767.94
joanne hayes-white    386807.48
amy hart              376071.59
Name: Pay, dtype: float64

###Highest paid males

In [ ]:
#Highest paid males
df[(df.Gender =="m")].groupby("EmployeeName").Pay.max().nlargest(5)


EmployeeName
nathaniel ford    567595.43
gary jimenez      538909.28
david shinn       510732.68
william j         436224.36
gregory p         425815.28
Name: Pay, dtype: float64

##The mean benefits paid

In [ ]:
#The mean benefits paid
#We got the mean for the benefits column using .mean() and assigned it to z
#You can then use format then round to round the values to two decimal places
z = df["Benefits"].mean()
format(round(z,2))

'18926.87'

###Mean benefits paid for males and females

In [ ]:
#Mean benefits paid for males and females 
#We can compare the benefits for males and females and the average
#We grouped by gender then got the mean for the benefits 
df.groupby("Gender",as_index=False).Benefits.mean()

Gender      Benefits
0      f  18022.397369
1      m  20301.137266

##Mean base pay

In [ ]:
#Mean base pay
#We got the mean for the base pay then we formatted it
x = df["BasePay"].mean()
format(round(x,2))

'66063.72'

###Mean base male and female pay

In [ ]:
#We grouped by gender then got the mean for the base pay
df.groupby("Gender",as_index=False).BasePay.mean()


Gender       BasePay
0      f  63052.231127
1      m  71740.294580

##Average pay per year for all employees

In [ ]:
#Average pay per year for all employees
#We grouped by year then we got the mean pay for each year
#In this case we used as_index false just to display in table form as opposed to normal
df.groupby(["Year"],as_index=False).Pay.mean()




Year            Pay
0  2011   71744.103871
1  2012  100583.308465
2  2013  101452.870647
3  2014  100270.347720

###Average pay per year for males

In [ ]:
#Average pay per year for males
#We filtered to get only for males
#We grouped by year then we got the mean pay for each year

df[(df.Gender =="m")].groupby(["Year"],as_index=False).Pay.mean()

Year            Pay
0  2011   81231.989677
1  2012  111840.148792
2  2013  112517.000832
3  2014  111040.083618

###Average pay per year for females

In [ ]:
#Average pay per year for males
#We filtered to get only for males
#We grouped by year then we got the mean pay for each year

df[(df.Gender =="f")].groupby(["Year"],as_index=False).Pay.mean()

Year           Pay
0  2011  64448.267951
1  2012  92325.481812
2  2013  94429.594821
3  2014  93700.354304

##Average female and male pay

In [ ]:
#Average male pay
#We grouped by gender then got the mean pay
df.groupby(["Gender"],as_index=False).Pay.mean()

Gender            Pay
0      f   86475.670283
1      m  104254.643859

In [ ]:
#We got the average male and female pay and assigned them a value
#We then got the average pay and assigned it a value
#We got the difference between the average pay and the average male and female pay
#We then made a comparison
#In the data report more comparisons are made
x = df[(df.Gender == "m")].Pay.mean()
y = df[(df.Gender == "f")].Pay.mean()
z = df.Pay.mean()
print("Men get "+str(x-z)+" more than the mean")
print("Women get "+str(z-y)+" less than the mean")


Men get 10547.873606587047 more than the mean
Women get 7231.099969851784 less than the mean


##Top 5 highest earning jobs

In [ ]:
#Top 5 earning jobs
#We grouped by job title. Then we got the mean pay
#We used nlargest then narrowed it to the top 5 

df.groupby("JobTitle").Pay.mean().nlargest(5)


JobTitle
chief investment officer                          436224.360000
chief, fire department                            408865.326667
general manager-metropolitan transit authority    399211.275000
gen mgr, public trnsp dept                        380696.440000
chief of police                                   375797.347500
Name: Pay, dtype: float64

###Top 5 highest earning jobs for men

In [ ]:
#Top 5 highest earning jobs for males
#We filtered gender to males
#We grouped by jobtitle then got the mean pay for each jobtitle
#We used nlargest and narrowed it to the top 5

df[(df.Gender=="m")].groupby("JobTitle").Pay.mean().nlargest(5)

JobTitle
chief investment officer                          436224.3600
general manager-metropolitan transit authority    399211.2750
gen mgr, public trnsp dept                        380696.4400
chief of police                                   375797.3475
dep chf of dept (fire dept)                       362562.2200
Name: Pay, dtype: float64

###Top 5 highest earning jobs for females

In [ ]:
#Top 5 highest earning jobs for females
#We filtered gender to females
#We grouped by jobtitle then got the mean pay for each jobtitle
#We used nlargest and narrowed it to the top 5

df[(df.Gender=="f")].groupby("JobTitle").Pay.mean().nlargest(5)

JobTitle
chief, fire department      408865.326667
assistant deputy chief 2    362306.210000
asst med examiner           352683.254286
dept head v                 349221.598182
adm, sfgh medical center    347079.706667
Name: Pay, dtype: float64

##Top 5 most popular jobs

In [ ]:
#Top 5 most popular jobs
#We got the job title then used value count to count the values
#Then used nlargest to get the top 5
df.JobTitle.value_counts().nlargest(5)



transit operator    9424
special nurse       5790
registered nurse    4955
custodian           3214
firefighter         3153
Name: JobTitle, dtype: int64

###Top 5 most popular male jobs

In [ ]:
#Top 5 most popular male jobs
#We filtered gender to male
#We got the job title column then used value count to count the number of values which are equal
#We used nlargest to get the top 5 

df[(df.Gender=="m")].JobTitle.value_counts().nlargest(5)

transit operator                5591
firefighter                     2346
deputy sheriff                  1888
police officer 3                1821
public svc aide-public works    1262
Name: JobTitle, dtype: int64

###Top 5 most popular female jobs

In [ ]:
#Top 5 most popular female jobs
#We filtered gender to male
#We got the job title column then used value count to count the number of values which are equal
#We used nlargest to get the top 5 

df[(df.Gender=="f")].JobTitle.value_counts().nlargest(5)

special nurse                3108
registered nurse             2592
transit operator             1876
recreation leader            1041
attorney (civil/criminal)     755
Name: JobTitle, dtype: int64

##Comparison between average male pay and average female pay.

In [ ]:
#Comparison between average male pay and average female pay.
#The female earnings as compared to male earnings in percentage

#We filtered gender to male then got the mean pay and assigned it to x
x = df[(df.Gender == "m")].Pay.mean()
#We filtered gender to female then got the mean pay and assigned to y
y = df[(df.Gender == "f")].Pay.mean()

(y/x)*100

82.94658835490928

##People with salaries above average

In [ ]:
#The total number of people with salaries above average
#We got the people with pay above mean pay
#In this case we used length to count the cells
len(df[df.Pay > df.Pay.mean()])

72991

###Males with salary above average

In [ ]:
#Number of males whose salary is above average
#We got the people with pay above mean pay 
#We used the ampersand sign to also get people whose gender is male
#We used length to count the cells
len(df[(df.Pay > df.Pay.mean()) & (df.Gender=='m')])

38667

###Females with salary above average

In [ ]:
#Number of females whose salary is above average
#We got the people with pay above mean pay 
#We used the ampersand sign to also get people whose gender is female
#We used length to count the cells
len(df[(df.Pay > df.Pay.mean()) & (df.Gender=='f')])


22438

## **Conclusion and Recommendations**

## **Conclusion**

From our analysis , we were able to deduce that the average pay per year for males is higher compared to the pay for females. In addition, the highest earning job in San Francisco  is that of the chief investment officer(CIO). 

The most popular job is that of the transit operator which is mostly populated by the male. The most popular female jobs  in San Francisco in the period of study worked as special nurses and registered nurses. 

The average pay for females was 86475.67USD and for males was 104254.64USD whereas the average pay for all was 93,707. This means that men earn 10547.87 more from mean pay and females earn 7231.09 less from the males. This means that females earn 82.94% of what men earn.

Among those earning above average pay, 52.98% were male and 30.74% were female whereas the rest were not specified their gender. 



### **Recommendations**

We recommend that since women earn 82% of what men earn, and only 30% of women earn above average pay, more study can be done to identify the reasons why women earn less than men in general.